Mathlib/Computability/AkraBazzi.lean

```diff
@@ -0,0 +1,1836 @@
+                  eventually_ge_atTop 0] with x hf hg hf' hg' hx_pos
+  intro u hu
+  have hbx : b * x ≤ x := calc
+    b * x ≤ 1 * x   := by gcongr; exact le_of_lt hb.2
+        _ = x       := by ring
+  have fx_nonneg : 0 ≤ f x := hf' x hbx
+  have gx_nonneg : 0 ≤ g x := hg' x hbx
+  refine ⟨?lb, ?ub⟩
+  case lb => calc
+    min c₁ c₃ * (f x + g x) = min c₁ c₃ * f x + min c₁ c₃ * g x     := by simp only [mul_add]
+      _ ≤ c₁ * f x + c₃ * g x := by
+              gcongr
+              · exact min_le_left _ _
+              · exact min_le_right _ _
+      _ ≤ f u + g u := by
+              gcongr
+              · exact (hf u hu).1
+              · exact (hg u hu).1
+  case ub => calc
+    max c₂ c₄ * (f x + g x) = max c₂ c₄ * f x + max c₂ c₄ * g x := by simp only [mul_add]
+      _ ≥ c₂ * f x + c₄ * g x := by gcongr
+                                    · exact le_max_left _ _
+                                    · exact le_max_right _ _
+      _ ≥ f u + g u := by gcongr
+                          · exact (hf u hu).2
+                          · exact (hg u hu).2
+
+protected lemma inv {f : ℝ → ℝ} (hf : GrowsPolynomially f) (hf' : ∀ᶠ x in atTop, 0 < f x) :
+    GrowsPolynomially fun x => (f x)⁻¹ := by
+  intro b hb
+  have hb_pos := hb.1
+  obtain ⟨c₁, ⟨hc₁_mem, ⟨c₂, ⟨hc₂_mem, hf⟩⟩⟩⟩ := hf b hb
+  have c₁_pos : 0 < c₁ := hc₁_mem
+  have c₂_pos : 0 < c₂ := hc₂_mem
+  refine ⟨c₂⁻¹, ⟨by show 0 < c₂⁻¹; positivity, ?_⟩⟩
+  refine ⟨c₁⁻¹, ⟨by show 0 < c₁⁻¹; positivity, ?_⟩⟩
+  filter_upwards [hf, hf', (tendsto_id.const_mul_atTop hb_pos).eventually_forall_ge_atTop hf']
+    with x hx hx' hx''
+  intro u hu
+  have h₁ : 0 < f u := hx'' u hu.1
+  refine ⟨?lb, ?ub⟩
+  case lb =>
+    rw [←mul_inv]
+    gcongr
+    exact (hx u hu).2
+  case ub =>
+    rw [←mul_inv]
+    gcongr
+    exact (hx u hu).1
+
+protected lemma div {f g : ℝ → ℝ} (hf : GrowsPolynomially f) (hg : GrowsPolynomially g)
+    (hf' : 0 ≤ᶠ[atTop] f) (hg' : ∀ᶠ x in atTop, 0 < g x) :
+    GrowsPolynomially fun x => f x / g x := by
+  have : (fun x => f x / g x) = fun x => f x * (g x)⁻¹ := by ext; rw [div_eq_mul_inv]
+  rw [this]
+  refine GrowsPolynomially.mul hf (GrowsPolynomially.inv hg hg') hf' ?_
+  filter_upwards [hg'] with x hx
+  positivity
+
+lemma congr_eventuallyEq {f g : ℝ → ℝ} (hfg : f =ᶠ[atTop] g) (hg : GrowsPolynomially g) :
+    GrowsPolynomially f := by
+  intro b hb
+  have hg' := hg b hb
+  obtain ⟨c₁, ⟨hc₁_mem, ⟨c₂, ⟨hc₂_mem, hg'⟩⟩⟩⟩ := hg'
+  refine ⟨c₁, ⟨hc₁_mem, ⟨c₂, ⟨hc₂_mem, ?_⟩⟩⟩⟩
+  filter_upwards [hg', (tendsto_id.const_mul_atTop hb.1).eventually_forall_ge_atTop hfg, hfg]
+    with x hx₁ hx₂ hx₃
+  intro u hu
+  rw [hx₂ u hu.1, hx₃]
+  exact hx₁ u hu
+
+protected lemma rpow (p : ℝ) : GrowsPolynomially fun x => x ^ p := by
+  intro b hb
+  have hb₀ : 0 < b := hb.1
+  have hbp₀ : 0 < b ^ p := Real.rpow_pos_of_pos hb₀ _
+  by_cases hp : 0 ≤ p
+  case pos =>
+    refine ⟨b^p, ⟨hbp₀, ?_⟩⟩
+    refine ⟨1, ⟨by norm_num, ?_⟩⟩
```

Comment:
normally we omit the nesting unless it really helps with symmetry:
```suggestion
    refine ⟨b^p, hbp₀, ?_⟩
    refine ⟨1, by norm_num, ?_⟩
```

Reply:
Strange -- I specifically remember this not working somehow!